### PR TITLE
Option to disable autocast when recording backward cuda graph

### LINKED ISSUE #120930: make_graphed_callables doesn't turn off autocast for backward
### 🐛 Describe the bug

The `make_graphed_callables` api for applying cuda graphs records both forward and backward kernels in cuda graphs. However, if `make_graphed_callables` is called within an autocast context to use AMP for the forward pass, it will also use autocast for the backward pass. [Pytorch docs](https://pytorch.org/docs/stable/amp.html#torch.autocast) recommends not using autocast for the backward pass. If users run backward outside of autocast during eager training, the ops can be different from the backward cuda graph recorded by `make_graphed_callables`.

Here's an example repro:
```Python
import torch
from torch.cuda.graphs import make_graphed_callables
from torch.testing import assert_close


class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, a, b):
        scale = 10
        return torch.bmm(a * scale, b * scale)


if __name__ == "__main__":
    baseline_in = (
        torch.randn(2, 576, 1536, device="cuda").requires_grad_(),
        torch.randn(2, 1536, 1536, device="cuda").requires_grad_(),
    )
    candidate_in = tuple(t.detach().clone().requires_grad_() for t in baseline_in)

    half_dtype = torch.bfloat16
    baseline = Repro()
    candidate = Repro()

    with torch.autocast(device_type="cuda", enabled=True, dtype=half_dtype, cache_enabled=False):
        make_graphed_callables(candidate, candidate_in)
    assert all(cc.grad is None for cc in candidate_in)

    with torch.autocast(device_type="cuda", enabled=True, dtype=half_dtype, cache_enabled=False):
        bloss = baseline(*baseline_in).mean()
    bloss.backward()

    with torch.autocast(device_type="cuda", enabled=True, dtype=half_dtype, cache_enabled=False):
        closs = candidate(*candidate_in).mean()
    closs.backward()

    for i in range(len(baseline_in)):
        assert_close(baseline_in[i].grad, candidate_in[i].grad)
```
The assert_close should fail. Interestingly, the assert_close passes if we change the autocast dtype to torch.float16.

Here is a similar issue: https://github.com/pytorch/pytorch/issues/71631

To fix the bug locally, I copied `make_graphed_callables` and disabled autocast when calling `autograd.grad`, and the repro script passes.

We could add this change to `make_graphed_callables`, or, at least provide an option to disable autocast for backward. Happy to make a PR if either option sounds good.

### Versions

Pytorch 2.1, H100 GPU

### PR DESCRIPTION
**Pull request by ErnestChan**
Fixes #120930

By default, `make_graphed_callables` will record the backward graph with autocast disabled.

cc @ezyang @gchanan @ptrblck @msaroufim @mcarilli @eellison @penguinwu @peterbell10

### CODE REVIEW

**Comment by ezyang** [comment]
Needs CLA.

Reading this implementation, I realized there is something a little dorky about this implementation: if you set the flag to True, we will turn on autocast on backwards, *even if there was no ambient autocast at all*. This should be incredibly unlikely to do anything! You actually want something like a contextstack, where you are going to push an autocast disable if the flag is False, and just don't do anything if the flag is True. (I also feel this phrasing is confusing and we should invert the boolean). WDYT.

File: torch/cuda/graphs.py [hunk 1]
```diff
@@ -187,7 +187,12 @@ def __exit__(self, exc_type, exc_value, traceback):
 
 
 def make_graphed_callables(
-    callables, sample_args, num_warmup_iters=3, allow_unused_input=False, pool=None
+    callables,
+    sample_args,
+    num_warmup_iters=3,
+    allow_unused_input=False,
+    pool=None,
+    enable_autocast_bwd=False,
```
Comment (eellison) [inline torch/cuda/graphs.py:195]: Yea, let's make this None, and only disable on `False`

Comment (ErnestChan) [reply torch/cuda/graphs.py:195]: Sounds good

**Comment by ErnestChan** [comment]
> Reading this implementation, I realized there is something a little dorky about this implementation: if you set the flag to True, we will turn on autocast on backwards, _even if there was no ambient autocast at all_. This should be incredibly unlikely to do anything! You actually want something like a contextstack, where you are going to push an autocast disable if the flag is False, and just don't do anything if the flag is True. (I also feel this phrasing is confusing and we should invert the boolean). WDYT.

Sounds good. Setting the flag to True if autocast is off would be weird and something we should prevent

File: torch/cuda/graphs.py [hunk 2]
```diff
@@ -187,7 +188,12 @@ def __exit__(self, exc_type, exc_value, traceback):
 
 
 def make_graphed_callables(
-    callables, sample_args, num_warmup_iters=3, allow_unused_input=False, pool=None
+    callables,
+    sample_args,
+    num_warmup_iters=3,
+    allow_unused_input=False,
+    pool=None,
+    disable_autocast_bwd=None,
```
Comment (ezyang) [inline torch/cuda/graphs.py:196]: Default this to True now, no?

File: torch/cuda/graphs.py [hunk 3]
```diff
@@ -348,19 +369,26 @@ def make_graphed_callables(
...
-
-        with torch.cuda.graph(bwd_graph, pool=mempool):
+        autocast_ctx = (
+            torch.autocast(
+                device_type=static_outputs[0].device.type,
+                enabled=False,
+                cache_enabled=False,
+            )
+            if disable_autocast_bwd and torch.is_autocast_enabled()
+            else nullcontext()
```
Comment (ezyang) [inline torch/cuda/graphs.py:389]: it's mildly annoying we have to dupe this code twice

**Review by ezyang — APPROVED** [review]
ok seems fine, please see review comments

File: torch/cuda/graphs.py (hunk 2 above)
Comment (ErnestChan) [reply torch/cuda/graphs.py:196]: I think @eellison wanted this to default to `None`?

File: torch/cuda/graphs.py (hunk 3 above)
Comment (ErnestChan) [reply torch/cuda/graphs.py]: Refactored to create the context manager with a helper function

**Comment by ezyang** [comment]
ci failure looks real

**Comment by ErnestChan** [comment]
hm I'm not quite sure how to fix:
```
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_SUPPORTED when calling `cublasGemmStridedBatchedEx(handle, opa, opb, (int)m, (int)n, (int)k, (void*)&falpha, a, CUDA_R_16BF, (int)lda, stridea, b, CUDA_R_16BF, (int)ldb, strideb, (void*)&fbeta, c, CUDA_R_16BF, (int)ldc, stridec, (int)num_batches, compute_type, CUBLAS_GEMM_DEFAULT_TENSOR_OP)`
```

Since I can't repro it on my machines

**Comment by ezyang** [comment]
If you change the default back to the legacy behavior, I expect that to be net neutral on the codebase, and then we can work out the CI failure later.

**Pull request closed without merging**